---
emoji: "🐥"
publication_name: "doctormate"
published: true
published_at: 2026-02-23 12:00
title: "axum crate のエラーハンドリング"
topics: ["rust"]
type: "tech"
---

[前回は axum crate の IntoResponse の標準で提供される実装のうち Result や Error などに関連するものを見ました](https://zenn.dev/doctormate/articles/093f5b5fb858b4)。

今回は axum のエラーハンドリングを見ていきます。

今回も axum crate のバージョンは 0.8.4 です。

## axum のエラーハンドリングの基礎

ドキュメントを見ていきます。

<https://docs.rs/axum/0.8.4/axum/error_handling/index.html>

冒頭から `tower::Service` の話で身構えてしまいそうですが、 `tower::Service` の関連型の `Error` をそのまま hyper crate まで到達させてしまうと応答を返さないことになってしまうので、 axum crate では `tower::Service` の関連型の `Error` に `Infallible` を取ることを要求することで回避しているということが書いてあります。エラー時にもエラーの応答を返さないとまずいですからね。

Handler の返す `Result<T, StatusCode>` の `StatusCode` はエラーではなく `IntoResponse` trait を実装しているので `Response` に変換されて返されるということも書いてあります。これはまさに[前回](https://zenn.dev/doctormate/articles/093f5b5fb858b4)の内容ですね。

examples へのリンクも示されています。

- <https://github.com/tokio-rs/axum/blob/axum-v0.8.4/examples/anyhow-error-response/src/main.rs>
    - `anyhow::Error` を wrap した `AppError` を定義し (newtype pattern) 、 `IntoResponse` を実装する例
    - `impl From<E> for AppError where E: Into<anyhow::Error>` によって `anyhow::Error` に変換できる任意のエラーから変換可能にしてある
- <https://github.com/tokio-rs/axum/blob/axum-v0.8.4/examples/error-handling/src/main.rs>
    - `enum AppError` を定義し、 `IntoResponse` を実装する例

Extractor でも同様に機能すると書いてあります。そうですね。 Extractor 、つまりは `FromRequest` も `FromRequestParts` も、関連型の `Rejection` が `IntoResponse` を要求していました。

- <https://docs.rs/axum/0.8.4/axum/extract/trait.FromRequest.html>
- <https://docs.rs/axum/0.8.4/axum/extract/trait.FromRequestParts.html>

このあたりは [Extractor の回](https://zenn.dev/doctormate/articles/10745f66b57301)に触れたように思います。

## HandleError

もうすこしドキュメントを読み進めます。 Service のエラーについてです。

<https://docs.rs/axum/0.8.4/axum/error_handling/index.html#routing-to-fallible-services>

Handler は Result を返しても `IntoResponse` であることが要求されるので問題ないです。そうではなく Service を設定するとエラーが問題になります。そこで `Error` を `Response` にする方法を設定するための `HandleError` の紹介が出てきます。

<https://docs.rs/axum/0.8.4/axum/error_handling/struct.HandleError.html>

ほぼ `IntoResponse` ですが、 `HandleError` の変換関数は `async fn` でも良いというのが面白いですね。

## HandleErrorLayer

同様に middleware のエラーについてです。

<https://docs.rs/axum/0.8.4/axum/error_handling/index.html#applying-fallible-middleware>

コード例はなんとなく分かりますが、 `ServiceBuilder` の前後関係などはそこまでピンと来ない感じがします。

<https://docs.rs/axum/0.8.4/axum/error_handling/struct.HandleErrorLayer.html>

実装を見ると分かりやすいのですが、結局のところ `HandleError` を返す `Layer` を提供してくれているだけです。

```rust
impl<S, F, T> Layer<S> for HandleErrorLayer<F, T>
where
    F: Clone,
{
    type Service = HandleError<S, F, T>;

    fn layer(&self, inner: S) -> Self::Service {
        HandleError::new(inner, self.f.clone())
    }
}
```

## HandleErrorLayer のための関数には Extractor を使える

しれっと書いてあります。

<https://docs.rs/axum/0.8.4/axum/error_handling/index.html#running-extractors-for-error-handling>

ここまで何度も見た T1 から T16 までの可変長に対応する実装ですね。

今回は `HandleError` の型の最後の Tuple 部分ですね。

```rust
impl<S, F, B, Res, Fut, $($ty,)*> Service<Request<B>>
    for HandleError<S, F, ($($ty,)*)>
```

## おわりに

今回は axum crate のエラーハンドリングとして Handler, Service, Middleware のエラーハンドリングと、そのために提供されている `HandleError` と `HandleErrorLayer` について書きました。
